header-rewrite-filter/header_processor.cc

```diff
@@ -533,6 +549,86 @@ namespace HeaderRewriteFilter {
     return std::make_tuple(absl::InvalidArgumentError("failed to execute dynamic function -- invalid function type"), "");
   }
 
+  absl::Status SetDynamicMetadataProcessor::parseOperation(std::vector<absl::string_view>& operation_expression, std::vector<absl::string_view>::iterator start) {
+    if (operation_expression.size() < Utility::SET_DYN_METADATA_MIN_NUM_ARGUMENTS) {
+        return absl::InvalidArgumentError("not enough arguments for set-metadata");
+    }
+
+    try {
+        absl::Status status;
+
+        metadata_key_ = std::string(*start);
+
+        absl::string_view value = *(start + 1);
+        metadata_value_ = std::make_shared<DynamicFunctionProcessor>(bool_processors_, is_request_);
+        status = metadata_value_->parseOperation(value);
+        if (status != absl::OkStatus()) {
+            return status;
+        }
+
+        if (start + 2 != operation_expression.end()) {
+            if (*(start + 2) == Utility::IF_KEYWORD) { // condition found
+                const absl::Status status = HeaderProcessor::ConditionProcessorSetup(operation_expression, start+3); // pass everything after the "if"
+                if (status != absl::OkStatus()) {
+                    return status;
+                }
+            } else {
+                return absl::InvalidArgumentError("third argument to set-metadata must be a condition");
+            }
+        }
+
+
+    } catch (const std::exception& e) {
+        // should never happen, range is checked above
+        return absl::InvalidArgumentError("error parsing set metadata operation -- " + std::string(e.what()));
+    }
+
+    return absl::OkStatus();
+  }
+
+  absl::Status SetDynamicMetadataProcessor::executeOperation(Http::RequestOrResponseHeaderMap& headers, Envoy::StreamInfo::StreamInfo& streamInfo) {
+    try {
+        const std::tuple<absl::Status, bool> condition_result = evaluateCondition(headers, streamInfo);
+        absl::Status status = std::get<0>(condition_result);
+        if (status != absl::OkStatus()) {
+            return status;
+        }
+
+        if (!std::get<1>(condition_result)) {
+            return absl::OkStatus(); // do nothing because condition is false
+        }
+
+        // get dynamic value to set
+        std::tuple<absl::Status, std::string> result = metadata_value_->executeOperation(headers, streamInfo);
+        status = std::get<0>(result);
+        const std::string value = std::get<1>(result);
+        if (status != absl::OkStatus() || value.length() == 0) {
+            return absl::InvalidArgumentError("failed to get dynamic value to set metadata " + std::string(status.message()));
```

Comment:
Nit: this would be easier to debug as 2 separate messages, e.g.
```
        if (status != absl::OkStatus()) {
            return absl::InvalidArgumentError("failed to get dynamic value to set metadata " + std::string(status.message()));
        } else if (value.length() == 0) {
            return absl::InvalidArgumentError("failed to get dynamic value to set metadata: no value");
        }
```